Implement a decorator that throttles an incoming request

imasahiro opened the pull request:
Motivations:
- A user somtimes want to throttle an incomming request to prevent a service
  from being overwhelmed by too many requests.

Modifications:
- Add RequestThrottling*

Results:
- Fixes https://github.com/line/armeria/issues/18

minwoox commented on core/src/main/java/com/linecorp/armeria/server/throttling/RequestThrottlingStrategy.java:
```diff
@@ -0,0 +1,59 @@
+import com.linecorp.armeria.common.Request;
+
+/**
+ * Determines whether a failed request should be retried.
```
a request should be throttled?

minwoox commented on core/src/main/java/com/linecorp/armeria/server/throttling/RequestThrottlingStrategy.java:
```diff
@@ -0,0 +1,59 @@
+     * Returns a {@link RequestThrottlingStrategy} that throttles {@code rate} percent of incoming requests.
+     */
+    static RequestThrottlingStrategy rateLimit(float rate, Supplier<Random> randomSupplier) {
+        return request -> randomSupplier.get().nextFloat() > rate;
```
requireNonNull(randomSupplier, "randomSupplier");

minwoox commented on core/src/main/java/com/linecorp/armeria/server/throttling/RequestThrottlingService.java:
```diff
@@ -0,0 +1,81 @@
+    public static <I extends Request, O extends Response>
+    Function<Service<I, O>, RequestThrottlingService<I, O>>
+    newDecorator(RequestThrottlingStrategy strategy) {
+        return newDecorator((ctx, req) -> strategy);
```
null check for the strategy?

minwoox commented on core/src/main/java/com/linecorp/armeria/server/throttling/RequestThrottlingService.java:
```diff
@@ -0,0 +1,81 @@
+    protected RequestThrottlingService(Service<I, O> delegate,
+                                       RequestThrottlingMapping mapping) {
+        super(delegate);
+        this.mapping = mapping;
```
null check for the mapping?

minwoox commented on core/src/main/java/com/linecorp/armeria/server/throttling/RequestThrottlingService.java:
```diff
@@ -0,0 +1,81 @@
+               .counter("armeria_request_throttling", Tags.zip("name", strategy.name(),
+                                                               "type", "throttled_count"))
+               .increment();
+            throw new HttpResponseException(HttpStatus.TOO_MANY_REQUESTS);
```
I am not very sure, but `TOO_MANY_REQUESTS` seems like used when "a client" send too many requests

trustin replied:
- How about adding a protected callback method with default behavior, just like `HttpAuthService`?
- How about using `SERVICE_UNAVAILABLE`?

minwoox commented on core/src/main/java/com/linecorp/armeria/server/throttling/RequestThrottlingMapping.java:
```diff
@@ -0,0 +1,45 @@
+     * per method.
+     */
+    static RequestThrottlingMapping methodStrategy(Map<String, RequestThrottlingStrategy> strategies) {
+        return (ctx, req) -> {
```
null check for the strategies?

trustin commented on core/src/main/java/com/linecorp/armeria/server/throttling/RequestThrottlingStrategy.java:
```diff
@@ -0,0 +1,59 @@
+    /**
+     * Returns whether a request should be throttled according to the given request.
+     */
+    boolean shouldThrottle(Request request);
```
- `CompletableFuture<Boolean>`?
- `request` could be parameterized?

trustin commented on core/src/main/java/com/linecorp/armeria/server/throttling/RequestThrottlingStrategy.java:
```diff
@@ -0,0 +1,59 @@
+     * Returns a name of this {@link RequestThrottlingStrategy}.
+     */
+    default String name() {
+        return "";
```
Auto-generate with `AtomicInteger.incrementAndGet()`?

trustin commented on core/src/main/java/com/linecorp/armeria/server/throttling/RequestThrottlingStrategy.java:
```diff
@@ -0,0 +1,59 @@
+    /**
+     * Returns a {@link RequestThrottlingStrategy} that throttles {@code rate} percent of incoming requests.
+     */
+    static RequestThrottlingStrategy rateLimit(float rate, Supplier<Random> randomSupplier) {
```
I'm not sure this will ever be used directly by a user. What a user wants would be to create a `RequestThrottlingStrategy` which can be turned on/off dynamically.

imasahiro replied:
Added sample impl for RequestThrottlingMapping that can be turend on/off dynamically based on file. WDYT?
https://github.com/line/armeria/pull/726/files#diff-7a7f8ccd97b41bee2ddf18eda232edaf

trustin commented on core/src/main/java/com/linecorp/armeria/server/throttling/RequestThrottlingService.java:
```diff
@@ -0,0 +1,81 @@
+        if (strategy.shouldThrottle(req)) {
+            ctx.meterRegistry()
+               .counter("armeria_request_throttling", Tags.zip("name", strategy.name(),
+                                                               "type", "throttled_count"))
```
```java
.counter(MeterRegistryUtil.name(registry, MeterUnit.NONE_TOTAL, "armeria", "server", "throttledRequests"),
         Tags.zip("name", strategy.name()))
```

Also, please keep the reference to the `Counter` so we don't keep looking it up. e.g. A weak map of strategy and `Counter`.

trustin commented on core/src/main/java/com/linecorp/armeria/server/throttling/RequestThrottlingStrategy.java:
```diff
@@ -0,0 +1,59 @@
+    /**
+     * A {@link RequestThrottlingStrategy} that defines a throttling should not be performed.
+     */
+    RequestThrottlingStrategy NEVER = request -> false;
```
- How about making this a method? `never()`
- How about adding `always()` as well?

trustin commented on core/src/main/java/com/linecorp/armeria/server/throttling/RequestThrottlingMapping.java:
```diff
@@ -0,0 +1,48 @@
+            return strategies.getOrDefault(method, RequestThrottlingStrategy.never());
+        };
+    }
+
```
How about adding the factory method for file-based one?

trustin commented on core/src/main/java/com/linecorp/armeria/server/throttling/RequestThrottlingMapping.java:
```diff
@@ -0,0 +1,48 @@
+ * Returns a {@link RequestThrottlingStrategy} instance from remote invocation parameters.
+ */
+@FunctionalInterface
+public interface RequestThrottlingMapping {
```
I'm curious why we need this interface. Couldn't we just add `ServiceRequestContext` to `RequestThrottlingStrategy.shouldThrottle(..)` and compose strategies?

trustin commented on core/src/main/java/com/linecorp/armeria/server/throttling/RequestThrottlingStrategy.java:
```diff
@@ -0,0 +1,54 @@
+ * Determines whether a request should be throttled.
+ */
+@FunctionalInterface
+public interface RequestThrottlingStrategy {
```
How about just `ThrottlingStrategy` just like `RetryStrategy`?

kojilin commented on core/src/main/java/com/linecorp/armeria/server/throttling/FileBasedDynamicRequestThrottlingMapping.java:
```diff
@@ -0,0 +1,132 @@
+                return RequestThrottlingStrategy.never();
+            case "rate":
+                float rate = Floats.tryParse(spec.get(1));
+                return req -> CompletableFuture.completedFuture(rate > ThreadLocalRandom.current().nextFloat());
```
static import?

kojilin commented on core/src/main/java/com/linecorp/armeria/server/throttling/FileBasedDynamicRequestThrottlingMapping.java:
```diff
@@ -0,0 +1,132 @@
+
+    @Override
+    public RequestThrottlingStrategy get(ServiceRequestContext ctx, Request req) {
+        String method = req instanceof RpcRequest ? ((RpcRequest) req).method() : ctx.method().name();
```
Can we do like path#method ? sometimes same method name may bind to different path. And I think it's good to using same pattern for http handler. e.g. `/api#POST`.

hyangtack commented on core/src/main/java/com/linecorp/armeria/server/throttling/ThrottlingService.java:
```diff
@@ -0,0 +1,92 @@
+    protected HttpResponse onFailure(ServiceRequestContext ctx, HttpRequest req, @Nullable Throwable cause)
+            throws Exception {
+        final DefaultHttpResponse res = new DefaultHttpResponse();
+        res.respond(HttpStatus.UNAUTHORIZED);
```
How about getting a throttled response status code from `ThrottlingStrategy`? A user might want to have their own throttling status code, so I think it would be better to be customizable.

hyangtack commented on core/src/main/java/com/linecorp/armeria/server/throttling/FileBasedDynamicThrottlingStrategy.java:
```diff
@@ -0,0 +1,150 @@
+    @Override
+    public CompletableFuture<Boolean> shouldThrottle(ServiceRequestContext ctx, Request req) {
+        String method = req instanceof RpcRequest ? ((RpcRequest) req).method() : ctx.method().name();
+        return strategies.getOrDefault(method, ThrottlingStrategy.never()).shouldThrottle(ctx, req);
```
Just curious. Is this enough to throttle a request based on the request method? What do you think about adding a simple path pattern as a key?

jmostella commented on core/src/main/java/com/linecorp/armeria/server/throttling/FileBasedDynamicThrottlingStrategy.java:
```diff
@@ -0,0 +1,157 @@
+                                                          StandardWatchEventKinds.ENTRY_CREATE,
+                                                          StandardWatchEventKinds.ENTRY_MODIFY
+                                                  },
+                                                  SensitivityWatchEventModifier.HIGH);
```
It looks like this will poll the file system every 2 seconds. Maybe this excessive? or should be configurable at least?

imasahiro commented:
Added TokenBucketThrottlingStrategy that throttoles the request based on current qps.
Also added CircuitBreakerThrottlingStrategy that throttoles the request based on existing CircuitBreaker

trustin commented on core/src/main/java/com/linecorp/armeria/server/throttling/ThrottlingStrategy.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    /**
+     * A {@link ThrottlingStrategy} that defines a throttling should be performed.
```
- A -> Returns a singleton
- that defines ... -> that fails all requests.
- Could we make this method return a singleton?

trustin commented on core/src/main/java/com/linecorp/armeria/server/throttling/ThrottlingStrategy.java:
```diff
@@ -0,0 +1,74 @@
+     * Returns a name of this {@link ThrottlingStrategy}.
+     */
+    public String name() {
+        return "strategy" + strategyId;
```
`"throttling-strategy-"`? (to be consistent with circuit breakers)

trustin commented on core/src/main/java/com/linecorp/armeria/server/throttling/ThrottlingStrategy.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    /**
+     * Returns whether a request should be throttled according to the given request.
```
How about: Returns whether a given request should be treated as failed before it is handled actually?

trustin commented on core/src/main/java/com/linecorp/armeria/server/throttling/ThrottlingStrategy.java:
```diff
@@ -0,0 +1,74 @@
+    /**
+     * Returns whether a request should be throttled according to the given request.
+     */
+    public abstract CompletableFuture<Boolean> shouldThrottle(ServiceRequestContext ctx, T request);
```
How about inverting the meaning of this method and renaming to `accept` or `canServe`? e.g. fail 20% of incoming requests vs accept 80% of incoming requests.

trustin commented on core/src/main/java/com/linecorp/armeria/server/throttling/TokenBucketThrottlingStrategy.java:
```diff
@@ -0,0 +1,71 @@
+
+/**
+ * A {@link ThrottlingStrategy} that throttles a request using
+ * {@see <a href="http://en.wikipedia.org/wiki/Token_bucket">Token Bucket</a>}.
```
`{@see}` is unnecessary?

trustin commented on core/src/main/java/com/linecorp/armeria/server/throttling/TokenBucketThrottlingStrategy.java:
```diff
@@ -0,0 +1,71 @@
+     * Creates a new strategy.
+     */
+    public TokenBucketThrottlingStrategy(int tokenBucketCapacity, int averageQPS, Duration refillInterval) {
+        checkArgument(tokenBucketCapacity > 0, "tokenBucketCapacity > 0");
```
`tokenBucketCapacity: %s (expected: > 0)", tokenBucketCapacity`

trustin commented on core/src/main/java/com/linecorp/armeria/server/throttling/TokenBucketThrottlingStrategy.java:
```diff
@@ -0,0 +1,71 @@
+     */
+    public TokenBucketThrottlingStrategy(int tokenBucketCapacity, int averageQPS, Duration refillInterval) {
+        checkArgument(tokenBucketCapacity > 0, "tokenBucketCapacity > 0");
+        checkArgument(averageQPS > 0, "averageQPS > 0");
```
`averageQPS: %s (expected: > 0)", averageQPS`

trustin commented on core/src/main/java/com/linecorp/armeria/server/throttling/TokenBucketThrottlingStrategy.java:
```diff
@@ -0,0 +1,71 @@
+            return completedFuture(true);
+        }
+
+        tokensInBucket--;
```
Thread safety? Probably better using `AtomicInteger` and CaS?

anuraaga commented on core/src/main/java/com/linecorp/armeria/server/throttling/FileBasedDynamicThrottlingStrategy.java:
```diff
@@ -0,0 +1,158 @@
+                        for (WatchEvent<?> event : key.pollEvents()) {
+                            // We only register "ENTRY_MODIFY" so the context is always a Path.
+                            final Path changed = (Path) event.context();
+                            if (changed.equals(settingFile.getFileName())) {
```
This comparison doesn't work when symlinks are involved (e.g., a Kubernetes config map where the specified human-readable filename like `settings.txt` will be a symlink to a timestamped uuid type filename like `settings.fq38rua3r908auf.txt`, and the event gets triggered with the timestamped file, not symlink file update.

I'm not sure if it's the best solution, but resolving the real paths during this check could help
https://github.com/curioswitch/curiostack/commit/0164c45c76df6897ed7bc5c7ebc2efe6da938272

imasahiro replied:
Nice catch! And thanks for reference.
To focus on other basic throttoring, I removed FileBasedDynamicThrottlingStrategy.java from this PR  strategies. Let me try dig more by myself on next PR.

anuraaga commented on core/src/main/java/com/linecorp/armeria/server/throttling/TokenBucketThrottlingStrategy.java:
```diff
@@ -0,0 +1,71 @@
+ * A {@link ThrottlingStrategy} that throttles a request using
+ * {@see <a href="http://en.wikipedia.org/wiki/Token_bucket">Token Bucket</a>}.
+ */
+public final class TokenBucketThrottlingStrategy<T extends Request> extends ThrottlingStrategy<T> {
```
You might be able to use Guava's `RateLimiter` and `return completedFuture(limiter.tryAcquire())`

trustin commented on core/src/main/java/com/linecorp/armeria/server/throttling/FileBasedDynamicThrottlingStrategy.java:
```diff
@@ -0,0 +1,174 @@
+            newStrategies = OBJECT_MAPPER.readValue(settingFile.toFile(),
+                                                    new TypeReference<Map<String, String>>() {});
+        } catch (IOException e) {
+            logger.warn("Could not read hosts file", e);
```
`hosts file -> the settings file: {}", settingFile, e` ?

imasahiro commented:
Pushed the changed but following FileBasedDynamicThrottlingStrategy.java comments are not reached yet.
https://github.com/line/armeria/pull/726#discussion_r143634419
https://github.com/line/armeria/pull/726#discussion_r143638201

> You may want to send a separate PR for strategy/mapping implementations because building a highly customizable strategy will make this PR take longer.

I feel, it better to split the PR like 1) add throttling feature with basic strategies, and 2) add other strategies like FileBasedDynamicThrottlingStrategy.. WDYT?

trustin commented:
> I feel, it better to split the PR like 1) add throttling feature with basic strategies, and 2) add other strategies like FileBasedDynamicThrottlingStrategy.. WDYT?

SGTM. Thanks for your patience!

anuraaga commented on core/src/main/java/com/linecorp/armeria/server/throttling/ThrottlingStrategy.java:
```diff
@@ -0,0 +1,78 @@
+     * Returns the name of this {@link ThrottlingStrategy}.
+     */
+    public String name() {
+        return "throttling-strategy-" + strategyId;
```
Not sure but this might work and provide more info

`getClass().isAnonymousClass ? strategyid : getClass().getName`

anuraaga commented on core/src/test/java/com/linecorp/armeria/server/throttling/RateLimitingThrottlingStrategyTest.java:
```diff
@@ -0,0 +1,56 @@
+                .isEqualTo(HttpStatus.OK);
+        client.get("/rate-limit").aggregate().get();
+        client.get("/rate-limit").aggregate().get();
+        // Reached to limit
```
This seems like it can be flaky? How about just using a unit test instead of server test?

anuraaga commented on core/src/test/java/com/linecorp/armeria/server/throttling/RateLimitingThrottlingStrategyTest.java:
```diff
@@ -0,0 +1,37 @@
+        assertThat(strategy.accept(null, null).get()).isEqualTo(false);
+        assertThat(strategy.accept(null, null).get()).isEqualTo(false);
+
+        TimeUnit.SECONDS.sleep(1);
```
I don't think this changes the flakiness much, we should add a `@VisibleForTesting` constructor that can take a mocked ratelimiter, or at least use awaitility here instead of a fixed sleep

imasahiro replied:
added VisibleForTesting constructor and inject mocked RateLimiter.

trustin commented on core/src/main/java/com/linecorp/armeria/server/throttling/RateLimitingThrottlingStrategy.java:
```diff
@@ -0,0 +1,69 @@
+     *
+     * @param requestPerSecond a number of requests per one second this {@link ThrottlingStrategy} accepts.
+     */
+    public RateLimitingThrottlingStrategy(int requestPerSecond, @Nullable String name) {
```
Should use `double`

trustin commented on core/src/main/java/com/linecorp/armeria/server/throttling/RateLimitingThrottlingStrategy.java:
```diff
@@ -0,0 +1,69 @@
+     *
+     * @param requestPerSecond a number of requests per one second this {@link ThrottlingStrategy} accepts.
+     */
+    public RateLimitingThrottlingStrategy(int requestPerSecond) {
```
Should use `double`

trustin commented:
Thanks, @imasahiro !

imasahiro commented:
Thanks a lot, all reviewers!
